logstash-core/src/main/java/org/logstash/instrument/metrics/ExtendedFlowMetric.java

```diff
@@ -0,0 +1,360 @@
+        if (possibleCapture.isEmpty()) { return Map.of(); }
+
+        final FlowCapture currentCapture = possibleCapture.get();
+
+        final Map<String, Double> rates = new LinkedHashMap<>();
+
+        this.retentionWindows.get()
+                             .forEach(window -> window.baseline(currentCapture.nanoTime())
+                                                      .or(() -> windowDefaultBaseline(window))
+                                                      .map((baseline) -> calculateRate(currentCapture, baseline))
+                                                      .orElseGet(OptionalDouble::empty)
+                                                      .ifPresent((rate) -> rates.put(window.policy.policyName(), rate)));
+
+        injectLifetime(currentCapture, rates);
+
+        return Collections.unmodifiableMap(rates);
+    }
+
+    /**
+     * Appends the given {@link FlowCapture} to the existing {@link RetentionWindow}s, XOR creates
+     * a new list of {@link RetentionWindow} using the provided {@link FlowCapture} as a baseline.
+     * This is concurrency-safe and uses non-volatile memory access in the hot path.
+     */
+    private void appendCapture(final FlowCapture capture) {
+        this.retentionWindows.ifSetOrElseSupply(
+                (existing) -> injectIntoRetentionWindows(existing, capture),
+                (        ) -> initRetentionWindows(retentionPolicies, capture)
+        );
+    }
+
+    private static List<RetentionWindow> initRetentionWindows(final Collection<? extends FlowMetricRetentionPolicy> retentionPolicies,
+                                                       final FlowCapture capture) {
+        return retentionPolicies.stream()
+                                .map((p) -> new RetentionWindow(p, capture))
+                                .collect(Collectors.toUnmodifiableList());
+    }
+
+    private static void injectIntoRetentionWindows(final List<RetentionWindow> retentionWindows, final FlowCapture capture) {
+        retentionWindows.forEach((rw) -> rw.append(capture));
+    }
+
+    /**
+     * Internal tooling to select the younger of two captures
+     */
+    private static FlowCapture selectNewerCapture(final FlowCapture existing, final FlowCapture proposed) {
+        if (existing == null) { return proposed; }
+        if (proposed == null) { return existing; }
+
+        return (existing.nanoTime() > proposed.nanoTime()) ? existing : proposed;
+    }
+
+    /**
+     * If a window's policy allows it to report before its retention has been reached,
+     * use our lifetime baseline as a default.
+     */
+    private Optional<FlowCapture> windowDefaultBaseline(final RetentionWindow window) {
+        if (window.policy.reportBeforeSatisfied()) {
+            return this.lifetimeBaseline.asOptional();
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Internal introspection for tests to measure how many captures we are retaining.
```

Comment:
Yay, another reason why we should onboard `guava` lib. [`@VisibleForTesting`](https://guava.dev/releases/19.0/api/docs/com/google/common/annotations/VisibleForTesting.html) would be perfect fit here.